### DISCUSSION TITLE
How to work with CLS inside a bull queue

### DISCUSSION THREAD

**Discussion by pandektes-braedon** [Papooch/nestjs-cls#178] [Q&A] [answered]
I'd like to access CLS context within a consumer class, i.e. a class decorated with [@Processor](https://docs.nestjs.com/techniques/queues#consumers). We are using `@nestjs/bull` and have a single consumer that handles multiple jobs within the consumer. I've so far been able make things work by decorating each method in my consumer with the `UseCls` decorator, e.g. 

```
  @UseCls({
    async setup(cls, job) {
      cls.set('userId', job.data.userId)
     // run some other setup
    },
  })
  @Process(NamedJob.INDEX)
  async handleIndex(job: Job<ProcessorJob>): Promise<void> {
      // job logic
  }
```

The setup I run in the UseCls decorator is the same for each method. Ideally, I'd like to avoid duplicating this everywhere. My initial idea was to use [applyDecorators](https://docs.nestjs.com/custom-decorators#decorator-composition) to compose the `Process` and `UseCls` decorators together, however, I encounter the following typer error on `UseCls` which suggests to me this isn't the correct way to compose these decorators or it simply isn't possible: 

```
import { Process as BullProcess } from '@nestjs/bull'
import { applyDecorators } from '@nestjs/common'
import { UseCls } from 'nestjs-cls'

export function ProcessWithCls(name: string) {
  return applyDecorators(
    BullProcess(name),
    UseCls({
      setup(cls, ...args) {
        // do some setup
      },
    }),
  )
}

// type error
Argument of type '(target: any, propertyKey: string | symbol, descriptor: TypedPropertyDescriptor<(...args: any) => Promise<any>>) => void' is not assignable to parameter of type 'ClassDecorator | MethodDecorator | PropertyDecorator'.
  Type '(target: any, propertyKey: string | symbol, descriptor: TypedPropertyDescriptor<(...args: any) => Promise<any>>) => void' is not assignable to type 'ClassDecorator'.
    Target signature provides too few arguments. Expected 3 or more, but got 1.
```

Am I missing anything obvious here that would enable me to reduce the amount of boilerplate I'm need to write to access CLS in queues? I'm new to nestjs, working with decorators, and using cls in general, so it's very likely I'm doing something wrong. My understanding is that since my usage is [outside of a web request](https://papooch.github.io/nestjs-cls/features-and-use-cases/usage-outside-of-web-request) my options here might be limited. 

Any input or ideas would be greatly appreciated!

**Comment by Papooch** [comment, marked as answer]
Hi, your idea is generally correct, and if you explicitly cast `UseCls` to `any`, then it should work as expected at runtime.
EDIT: The recommended solution has [been added to the docs](https://papooch.github.io/nestjs-cls/setting-up-cls-context/using-a-decorator#gotchas).

The issue here is that the `UseCls` decorator doesn't, for some reason, satisfy the general decorator interfaces that Nest's method expects. Likely becaue the `UseCls` decorator is generic and uses a bit of typescript magic to be type-safe according to the function on which you use it.

I'll try to investigate this more on my end and get back with other possible solutions.

Btw, if you don't want to repeat a decorator on each method of a class, you might try my `decorate-all` npm package, although you might bump into the same typing issue.

**Reply by pandektes-braedon** [reply]
Great thanks for the feedback and I appreciate your work on investigating the typing further. 

Will take a look at the `decorate-all` pacakge!

**Reply by Papooch** [reply]
@pandektes-braedon I have updated the answer with a link to a new section in the docs.
